File: parser/status.py
"""Статус прогонов пайплайна → Supabase (таблица pipeline_runs).

Питает админ-дешборд «последнее обновление / успех / ошибки». Пишется в конце
скрейпа и НИКОГДА не должен ронять сам сбор: без Supabase-env — тихий no-op,
любая ошибка проглатывается.
"""
import os
import json
import urllib.request
import urllib.error

_URL = (os.environ.get("SUPABASE_URL") or "").rstrip("/")
_KEY = os.environ.get("SUPABASE_SERVICE_ROLE_KEY") or ""


def _enabled():
    return bool(_URL and _KEY)


def _headers(extra=None):
    h = {"apikey": _KEY, "Authorization": f"Bearer {_KEY}", "Content-Type": "application/json"}
    if extra:
        h.update(extra)
    return h


def prev_total(workflow):
    """Total прошлого прогона этого workflow (для health-check) или None."""
    if not _enabled():
        return None
    try:
        url = (f"{_URL}/rest/v1/pipeline_runs?workflow=eq.{workflow}"
               f"&select=total&order=finished_at.desc&limit=1")
        req = urllib.request.Request(url, headers=_headers())
        with urllib.request.urlopen(req, timeout=15) as r:
            rows = json.loads(r.read().decode())
        return rows[0]["total"] if rows else None
    except Exception:  # noqa: BLE001
        return None


def record(workflow, *, total, failed=0, prev=None, error=None, meta=None):
    """Пишет строку прогона. ok вычисляется: нет фатальной ошибки и нет резкого
    (>50%) падения total относительно прошлого прогона."""
    if not _enabled():
        print(f"[status] Supabase не сконфигурирован — пропуск ({workflow})")
        return
    ok = error is None
    warn = None
    if prev and total is not None and total < prev * 0.5:
        ok, warn = False, f"резкое падение: было {prev}, стало {total}"
    row = {
        "workflow": workflow, "ok": ok, "total": total, "prev_total": prev,
        "failed": failed, "error": error or warn, "meta": meta,
    }
    try:
        req = urllib.request.Request(
            f"{_URL}/rest/v1/pipeline_runs",
            data=json.dumps([row]).encode(),
            headers=_headers({"Prefer": "return=minimal"}), method="POST")
        urllib.request.urlopen(req, timeout=15).read()
        print(f"[status] записан прогон {workflow}: ok={ok} total={total} failed={failed}")
    except Exception as e:  # noqa: BLE001
        print(f"[status] не удалось записать прогон: {e}")

File: api/status.js
// GET /api/status — состояние пайплайнов сбора (только для роли admin).
// Отдаёт последний прогон каждого workflow + недавнюю историю с health-флагами.
import { userFromReq, sb, sbConfig } from "./_auth.js";

export default async function handler(req, res) {
  const u = userFromReq(req);
  if (!u) return res.status(401).json({ error: "не авторизован" });
  if (u.role !== "admin") return res.status(403).json({ error: "нужны права администратора" });
  if (!sbConfig().ok) return res.status(500).json({ error: "сервер не сконфигурирован (SUPABASE env)" });

  // Последние 100 прогонов, свежие сверху.
  const r = await sb("pipeline_runs?select=*&order=finished_at.desc&limit=100");
  if (!r.ok) return res.status(502).json({ error: "ошибка БД" });
  const runs = await r.json();

  // Последний прогон по каждому workflow — это и есть «карточки статуса».
  const latest = {};
  for (const run of runs) if (!latest[run.workflow]) latest[run.workflow] = run;

  return res.status(200).json({ latest: Object.values(latest), history: runs, now: new Date().toISOString() });
}
